Userland/Test.cpp

```diff
@@ -0,0 +1,534 @@
+        } else if (arg == "-gt") {
+            StringView rhs = argv[++optind];
+            return make<NumericCompare>(lhs, rhs, NumericCompare::Greater);
+        } else if (arg == "-le") {
+            StringView rhs = argv[++optind];
+            return make<NumericCompare>(lhs, rhs, NumericCompare::LessOrEqual);
+        } else if (arg == "-lt") {
+            StringView rhs = argv[++optind];
+            return make<NumericCompare>(lhs, rhs, NumericCompare::Less);
+        } else if (arg == "-ne") {
+            StringView rhs = argv[++optind];
+            return make<NumericCompare>(lhs, rhs, NumericCompare::NotEqual);
+        } else if (arg == "-ef") {
+            StringView rhs = argv[++optind];
+            return make<FileCompare>(lhs, rhs, FileCompare::Same);
+        } else if (arg == "-nt") {
+            StringView rhs = argv[++optind];
+            return make<FileCompare>(lhs, rhs, FileCompare::ModificationTimestampGreater);
+        } else if (arg == "-ot") {
+            StringView rhs = argv[++optind];
+            return make<FileCompare>(lhs, rhs, FileCompare::ModificationTimestampLess);
+        } else if (arg.is_null()) {
+            return make<StringCompare>("", lhs, StringCompare::NotEqual);
+        } else if (g_was_invoked_as_bracket && arg == "]") {
+            // Ignore everything after the closing bracket in '[ ... ] foo bar'.
+            return nullptr;
+        }
+    }
+
+    fatal_error("Unsupported option \033[1m%s", argv[optind]);
+}
+
+OwnPtr<Condition> parse_complex_expression(char* argv[])
+{
+    auto command = parse_simple_expression(argv);
+
+    while (command && argv[optind] && argv[optind + 1]) {
+        StringView arg = argv[++optind];
+
+        // Ignore everything after the closing bracket in '[ ... ] foo bar'.
+        if (g_was_invoked_as_bracket && arg == "]")
+            return command;
+
+        enum {
+            AndOp,
+            OrOp,
+        } binary_operation { AndOp };
+
+        if (arg == "-a") {
+            optind++;
+            binary_operation = AndOp;
+        } else if (arg == "-o") {
+            optind++;
+            binary_operation = OrOp;
+        } else {
+            // Ooops, looked too far.
+            optind--;
+            return command;
+        }
+        auto rhs = parse_complex_expression(argv);
+        if (!rhs)
+            fatal_error("Missing right-hand side");
+
+        if (binary_operation == AndOp)
+            command = make<And>(command.release_nonnull(), rhs.release_nonnull());
+        else
+            command = make<Or>(command.release_nonnull(), rhs.release_nonnull());
+    }
+
+    if (g_was_invoked_as_bracket && StringView { argv[optind] } != "]")
+        fatal_error("test invoked as '[' requires a closing bracket ']'");
+
+    return command;
+}
+
+int main(int, char* argv[])
+{
+    if (pledge("stdio rpath", nullptr) < 0) {
+        perror("pledge");
+        return 126;
+    }
+
+    if (StringView { argv[0] } == "[")
```

Comment:
We should use this instead check instead.

This way `/bin/[ 1 ]` works as expected

```suggestion
    StringView arg0(argv[0]);
    if (arg0 == "[" || arg0.ends_with("/["))
```

Reply:
Is there a function to get the filename out of a path?

If there is, we could also write this as:
```cpp
if (function_that_gets_filename_from_path(StringView(argv[0])) == "[")
```

Reply:
https://github.com/SerenityOS/serenity/blob/master/Base/usr/share/man/man3/basename.md, or a similar thing in `AK::LexicalPath`

Reply:
I guess we can use `LexicalPath{argv[0]}.basename()`?

Reply:
Yep